File: tracemail-app/src/api/cases.ts
import { request, upload } from './client';
import type { CampaignData, Case, CopilotMessage, EvidenceFinding, EvidenceRelationship, ForensicAttachment, ForensicSummary, IdentityAnalysis, InfrastructureData, IOC, MimePart, ReportData, TimelineEvent, TransitHop } from '../types';

type WireCase = {
  id: string; number: number; title: string; subject: string; threat_type: Case['threatType']; risk_score: number;
  risk_level: Case['riskLevel']; identity_consistency: number; model_confidence: number; evidence_confidence: number;
  status: Case['status']; received_date: string; sender_name: string; sender_email: string; sender_domain: string;
  to_email: string; reply_to: string; return_path: string; received_ip: string; spf_result: Case['spfResult'];
  dkim_result: Case['dkimResult']; dmarc_result: Case['dmarcResult']; sha256: string; file_name: string; file_size: number;
};
const caseFromWire = (c: WireCase): Case => ({
  id: c.id, number: c.number, title: c.title, subject: c.subject, threatType: c.threat_type, riskScore: c.risk_score,
  riskLevel: c.risk_level, identityConsistency: c.identity_consistency, modelConfidence: c.model_confidence,
  evidenceConfidence: c.evidence_confidence, status: c.status, receivedDate: c.received_date, senderName: c.sender_name,
  senderEmail: c.sender_email, senderDomain: c.sender_domain, toEmail: c.to_email, replyTo: c.reply_to,
  returnPath: c.return_path, receivedIp: c.received_ip, spfResult: c.spf_result, dkimResult: c.dkim_result,
  dmarcResult: c.dmarc_result, sha256: c.sha256, fileName: c.file_name, fileSize: c.file_size, sourceType: (c as any).source_type, analysisStatus: (c as any).analysis_status,
});

export type UploadResult = { case_id: string; case_number: number; status: string; source_type: string; evidence_id: string; analysis_status: string; message: string };
export const listCases = async () => (await request<WireCase[]>('/api/cases')).map(caseFromWire);
export const getCase = async (id: string) => caseFromWire(await request<WireCase>(`/api/cases/${id}`));
export const uploadEmail = (file: File) => upload<UploadResult>('/api/cases/upload', file);
export const getIdentity = async (id: string): Promise<IdentityAnalysis> => {
  const value: any = await request(`/api/cases/${id}/identity`);
  return { caseId: value.case_id, score: value.score, displayName: value.display_name, fromAddress: value.from_address, fromDomain: value.from_domain, replyTo: value.reply_to, replyToDomain: value.reply_to_domain, returnPath: value.return_path, returnPathDomain: value.return_path_domain, dkimDomain: value.dkim_domain, sourceIp: value.source_ip, sourceAsn: value.source_asn, sourceLocation: value.source_location, sourceProvider: value.source_provider, contradictions: value.contradictions, chain: value.chain };
};
export const getEvidence = async (id: string): Promise<EvidenceFinding[]> => (await request<any[]>(`/api/cases/${id}/evidence`)).map(value => ({ id: value.id, title: value.title, description: value.description, severity: value.severity, riskBadge: value.risk_badge, code: value.code, evidenceState: value.evidence_state, confidence: value.confidence, source: value.source, tags: value.tags, colorAccent: value.color_accent }));
export const getHeaders = (id: string) => request<{ headers: Array<{key: string; value: string; is_anomalous?: boolean; badge?: string}>; raw: string }>(`/api/cases/${id}/headers`);
export const getTransitHops = async (id: string): Promise<TransitHop[]> => {
  const value: any = await request(`/api/cases/${id}/transit-hops`);
  return (value.hops ?? []).map((h: any) => ({
    label: h.label, ip: h.ip, description: h.description, type: h.hop_type,
    sequence: h.sequence, rawHeader: h.raw_header, fromHost: h.from_host,
    byHost: h.by_host, sourceIp: h.source_ip, protocol: h.protocol, timestamp: h.timestamp,
  }));
};
export const getMimeParts = async (id: string): Promise<MimePart[]> => (await request<any[]>(`/api/cases/${id}/mime-parts`)).map(v => ({ id: v.id, partIndex: v.part_index, parentId: v.parent_id, depth: v.depth, contentType: v.content_type, disposition: v.disposition, filename: v.filename, contentId: v.content_id, size: v.size, transferEncoding: v.transfer_encoding, isAttachment: v.is_attachment }));
export const getAttachments = async (id: string): Promise<ForensicAttachment[]> => (await request<any[]>(`/api/cases/${id}/attachments`)).map(v => ({ id: v.id, filename: v.filename, contentType: v.content_type, size: v.size, disposition: v.disposition, contentId: v.content_id, sha256: v.sha256 }));
export const getIOCs = async (id: string): Promise<IOC[]> => (await request<any[]>(`/api/cases/${id}/iocs`)).map(v => ({ id: v.id, type: v.type, value: v.value, source: v.source }));
export const getRelationships = async (id: string): Promise<EvidenceRelationship[]> => (await request<any[]>(`/api/cases/${id}/relationships`)).map(v => ({ id: v.id, relationType: v.relation_type, sourceType: v.source_type, sourceId: v.source_id, targetType: v.target_type, targetId: v.target_id }));
export const getForensicSummary = async (id: string): Promise<ForensicSummary> => { const v: any = await request(`/api/cases/${id}/forensic-summary`); return { caseId: v.case_id, evidenceId: v.evidence_id, sourceType: v.source_type, sourceIdentifier: v.source_identifier, acquisitionMethod: v.acquisition_method, acquisitionTimestamp: v.acquisition_timestamp, contentType: v.content_type, originalFilename: v.original_filename, size: v.size, sha256: v.sha256, mimeParts: v.mime_parts, receivedHops: v.received_hops, attachments: v.attachments, iocs: v.iocs }; };
export const getInfrastructure = async (id: string): Promise<InfrastructureData> => { const value: any = await request(`/api/cases/${id}/infrastructure`); return { caseId: value.case_id, sourceIp: value.source_ip, asn: value.asn, provider: value.provider, location: value.location, country: value.country, latitude: value.latitude, longitude: value.longitude, domain: value.domain, relatedDomains: value.related_domains, relatedCases: value.related_cases, threatScore: value.threat_score, infrastructureType: value.infrastructure_type, threatFeeds: value.threat_feeds }; };
export const getGraph = (id: string) => request<{ nodes: unknown[]; edges: unknown[] }>(`/api/cases/${id}/graph`);
export const getCampaign = async (id: string): Promise<CampaignData | null> => { const value: any = await request(`/api/cases/${id}/campaign`); return value && { id: value.id, name: value.name, relatedEmails: value.related_emails, sharedDomains: value.shared_domains, sharedInfrastructure: value.shared_infrastructure, relatedCases: value.related_cases, timeline: value.timeline.map((x: any) => ({date: x.date, title: x.title, type: x.event_type, description: x.description})), description: value.description }; };
export const getTimeline = async (id: string): Promise<TimelineEvent[]> => (await request<any[]>(`/api/cases/${id}/timeline`)).map(value => ({ id: value.id, time: value.time, title: value.title, description: value.description, type: value.event_type, icon: value.icon, details: value.details }));
const reportFromWire = (value: any): ReportData => ({caseId: value.case_id, executiveSummary: value.executive_summary, evidenceAnalysis: value.evidence_analysis, identityAnalysis: value.identity_analysis, infrastructureAnalysis: value.infrastructure_analysis, campaignCorrelation: value.campaign_correlation, attackPath: value.attack_path, confidenceAndLimitations: value.confidence_and_limitations, conclusion: value.conclusion, generatedAt: value.generated_at});
export const getReport = async (id: string) => reportFromWire(await request(`/api/reports/${id}`));
export const generateReport = async (id: string) => reportFromWire(await request(`/api/reports/${id}/generate`, { method: 'POST' }));
const messageFromWire = (value: any): CopilotMessage => ({id: value.id, role: value.role, content: value.content, timestamp: value.timestamp});
export const getCopilotMessages = async (id: string) => (await request<any[]>(`/api/copilot/${id}/messages`)).map(messageFromWire);
export const askCopilot = async (id: string, question: string) => messageFromWire(await request(`/api/copilot/${id}/ask`, { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify({question}) }));
export const getSuggestedQuestions = (id: string) => request<{questions: string[]}>(`/api/copilot/${id}/suggested-questions`, { method: 'POST' });

File: backend/routers/cases.py
"""Cases CRUD + .eml upload + full analysis orchestration."""
from __future__ import annotations
import json
from typing import Optional
from fastapi import APIRouter, Depends, File, UploadFile, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from database import get_db
from models import (
    Case, IdentityRecord, EvidenceFindingRecord, EvidenceRecord,
    InfrastructureRecord, TimelineEventRecord,
)
from schemas import (
    CaseResponse, UploadResponse, IdentityResponse,
    EvidenceFindingResponse, EmailHeaderOut, TransitHopOut,
    InfrastructureResponse, GraphNodeOut, GraphEdgeOut,
    CampaignResponse, CampaignTimelineEventOut,
    TimelineEventResponse, MimePartOut, AttachmentOut, IocOut,
    EvidenceRelationshipOut, ForensicSummary,
)
from services import evidence_queries
from services.sources import EMLSource
from services.email_parser import extract_received_hops_from_text
from services.ingest_email import ingest_acquired_email, UnparsableEmail
import json as _json

router = APIRouter(prefix="/api/cases", tags=["cases"])
MAX_EML_SIZE_BYTES = 10 * 1024 * 1024


@router.get("", response_model=list[CaseResponse])
async def list_cases(
    status: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
):
    q = select(Case).order_by(Case.number.desc())
    if status:
        q = q.where(Case.status == status)
    result = await db.execute(q)
    return result.scalars().all()


@router.get("/{case_id}", response_model=CaseResponse)
async def get_case(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Case).where(Case.id == case_id))
    case = result.scalar_one_or_none()
    if not case:
        raise HTTPException(404, "Case not found")
    return case


@router.post("/upload", response_model=UploadResponse)
async def upload_eml(
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
):
    if not file.filename or not file.filename.lower().endswith(".eml"):
        raise HTTPException(400, "Please upload a valid .eml file")

    raw_bytes = await file.read()
    if not raw_bytes:
        raise HTTPException(400, "The uploaded .eml file is empty")
    if len(raw_bytes) > MAX_EML_SIZE_BYTES:
        raise HTTPException(400, "The uploaded .eml exceeds the 10 MB limit")

    # Source capture: wrap the raw bytes + file metadata in the normalized
    # EmailSource contract, then hand the normalized acquisition to the
    # source-agnostic forensic pipeline.
    acquired = EMLSource(
        filename=file.filename,
        raw_bytes=raw_bytes,
        content_type=file.content_type or "message/rfc822",
    ).acquire()

    try:
        result = await ingest_acquired_email(db, acquired)
    except UnparsableEmail as exc:
        raise HTTPException(422, str(exc)) from exc

    return UploadResponse(
        case_id=result["case_id"],
        case_number=result["case_number"],
        status="active",
        source_type=acquired.source_type,
        evidence_id=result["evidence_id"],
        analysis_status="completed",
        message=(
            f"Analysis complete. Threat: {result['threat_type']}, "
            f"Risk: {result['risk_level']} ({result['risk_score']}/100)"
        ),
    )


# ─── Analysis Endpoints ───────────────────────────────────────

@router.get("/{case_id}/identity", response_model=IdentityResponse)
async def get_identity(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(IdentityRecord).where(IdentityRecord.case_id == case_id))
    identity = result.scalar_one_or_none()
    if not identity:
        raise HTTPException(404, "Identity analysis not found")
    return IdentityResponse(
        case_id=identity.case_id,
        score=identity.score,
        display_name=identity.display_name,
        from_address=identity.from_address,
        from_domain=identity.from_domain,
        reply_to=identity.reply_to,
        reply_to_domain=identity.reply_to_domain,
        return_path=identity.return_path,
        return_path_domain=identity.return_path_domain,
        dkim_domain=identity.dkim_domain,
        source_ip=identity.source_ip,
        source_asn=identity.source_asn,
        source_location=identity.source_location,
        source_provider=identity.source_provider,
        contradictions=[
            IdentityResponse.__annotations__["contradictions"]  # type: ignore
        ] if False else _parse_json_list(identity.contradictions_json),
        chain=_parse_json_list(identity.chain_json),
    )


@router.get("/{case_id}/evidence", response_model=list[EvidenceFindingResponse])
async def get_evidence(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(EvidenceFindingRecord).where(EvidenceFindingRecord.case_id == case_id)
    )
    findings = result.scalars().all()
    return [
        EvidenceFindingResponse(
            id=f.finding_id,
            title=f.title,
            description=f.description,
            severity=f.severity,
            risk_badge=f.risk_badge,
            code=f.code,
            evidence_state=f.evidence_state,
            confidence=f.confidence,
            source=f.source,
            tags=_json.loads(f.tags_json or "[]"),
            color_accent=f.color_accent,
        )
        for f in findings
    ]


# ─── Forensic evidence read APIs ───────────────────────────────
# Read-only views over the persisted normalized forensic records.
# These values are observables and metadata; none of it implies a verdict.

@router.get("/{case_id}/mime-parts", response_model=list[MimePartOut])
async def get_mime_parts(case_id: str, db: AsyncSession = Depends(get_db)):
    evidence = await _require_evidence(db, case_id)
    if evidence is None:
        return []
    parts = await evidence_queries.get_mime_parts(db, case_id)
    return [
        MimePartOut(
            id=p.id,
            part_index=p.part_index,
            parent_id=p.parent_index,
            depth=p.depth,
            content_type=p.content_type,
            disposition=p.disposition,
            filename=p.filename,
            content_id=p.content_id,
            size=p.size,
            transfer_encoding=p.transfer_encoding,
            is_attachment=p.is_attachment,
        )
        for p in parts
    ]


@router.get("/{case_id}/attachments", response_model=list[AttachmentOut])
async def get_attachments(case_id: str, db: AsyncSession = Depends(get_db)):
    evidence = await _require_evidence(db, case_id)
    if evidence is None:
        return []
    attachments = await evidence_queries.get_attachments(db, case_id)
    return [
        AttachmentOut(
            id=a.id,
            filename=a.filename,
            content_type=a.mime_type,
            size=a.size,
            disposition=a.disposition,
            content_id=a.content_id,
            sha256=a.sha256,
        )
        for a in attachments
    ]


@router.get("/{case_id}/iocs", response_model=list[IocOut])
async def get_iocs(case_id: str, db: AsyncSession = Depends(get_db)):
    evidence = await _require_evidence(db, case_id)
    if evidence is None:
        return []
    iocs = await evidence_queries.get_iocs(db, case_id)
    # An IOC is an observable: type, value and where it was seen. No IOC is
    # labelled malicious by the extraction process.
    return [IocOut(id=i.id, type=i.ioc_type, value=i.value, source=i.source) for i in iocs]


@router.get("/{case_id}/relationships", response_model=list[EvidenceRelationshipOut])
async def get_relationships(case_id: str, db: AsyncSession = Depends(get_db)):
    evidence = await _require_evidence(db, case_id)
    if evidence is None:
        return []
    relationships = await evidence_queries.get_relationships(db, case_id)
    return [
        EvidenceRelationshipOut(
            id=r.id,
            relation_type=r.relation_type,
            source_type=evidence.evidence_type,
            source_id=evidence.id,
            target_type=r.target_type,
            target_id=r.target_id,
        )
        for r in relationships
    ]


@router.get("/{case_id}/forensic-summary", response_model=ForensicSummary)
async def get_forensic_summary(case_id: str, db: AsyncSession = Depends(get_db)):
    summary = await evidence_queries.get_evidence_summary(db, case_id)
    if not summary.get("evidence_id"):
        return ForensicSummary(
            case_id=case_id, evidence_id="", source_type="", source_identifier="",
            acquisition_method="", acquisition_timestamp=None, content_type="",
            original_filename="", size=0, sha256="",
            mime_parts=0, received_hops=0, attachments=0, iocs=0,
        )
    return ForensicSummary(
        case_id=case_id,
        evidence_id=summary["evidence_id"],
        source_type=summary["source_type"],
        source_identifier=summary["source_identifier"],
        acquisition_method=summary["acquisition_method"],
        acquisition_timestamp=summary["acquisition_timestamp"],
        content_type=summary["content_type"],
        original_filename=summary["original_filename"],
        size=summary["size"],
        sha256=summary["sha256"],
        mime_parts=summary["mime_parts"],
        received_hops=summary["received_hops"],
        attachments=summary["attachments"],
        iocs=summary["iocs"],
    )


@router.get("/{case_id}/headers")
async def get_headers(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Case).where(Case.id == case_id))
    case = result.scalar_one_or_none()
    if not case:
        raise HTTPException(404, "Case not found")

    raw_headers = case.raw_headers or ""
    headers = []
    for line in raw_headers.split("\n"):
        line = line.strip()
        if not line:
            continue
        if ":" in line:
            key, _, value = line.partition(":")
            is_anomalous = key.strip() in ("Reply-To", "X-Mailer")
            badge = None
            if key.strip() == "Reply-To" and case.reply_to:
                badge = "Mismatch Detected"
            elif key.strip() == "From":
                badge = "Spoofed Brand" if case.sender_domain != "company.com" else None
            headers.append({
                "key": f"{key.strip()}:",
                "value": value.strip(),
                "is_anomalous": is_anomalous,
                "badge": badge,
            })
    return {"headers": headers, "raw": raw_headers}


@router.get("/{case_id}/transit-hops")
async def get_transit_hops(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Case).where(Case.id == case_id))
    case = result.scalar_one_or_none()
    if not case:
        raise HTTPException(404, "Case not found")

    # Phase-2+: return persisted forensic hop records (evidence-derived).
    persisted = await evidence_queries.get_received_hops(db, case_id)
    if persisted:
        return {"hops": [
            {
                "label": h.label,
                "ip": h.source_ip or h.by_host,
                "description": h.description,
                "hop_type": h.hop_type,
                "sequence": h.sequence,
                "raw_header": h.raw_header,
                "from_host": h.from_host,
                "by_host": h.by_host,
                "source_ip": h.source_ip,
                "protocol": h.protocol,
                "timestamp": h.timestamp,
            }
            for h in persisted
        ]}

    # Fallback for cases ingested before forensic records were persisted.
    # Hops are derived strictly from the Received headers in the actual EML.
    # No hostnames, IPs, organizations, or locations are invented; if the EML
    # contains no Received headers, the hop list is empty rather than fabricated.
    ev_result = await db.execute(
        select(EvidenceRecord).where(EvidenceRecord.case_id == case_id)
    )
    evidence = ev_result.scalars().first()
    eml_text = ""
    if evidence is not None and evidence.original_bytes:
        eml_text = evidence.original_bytes.decode("utf-8", errors="replace")
    if not eml_text:
        eml_text = case.raw_eml or ""
    return {"hops": extract_received_hops_from_text(eml_text)}


@router.get("/{case_id}/infrastructure", response_model=InfrastructureResponse)
async def get_infrastructure(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(InfrastructureRecord).where(InfrastructureRecord.case_id == case_id))
    infra = result.scalar_one_or_none()
    if not infra:
        raise HTTPException(404, "Infrastructure data not found")
    return InfrastructureResponse(
        case_id=infra.case_id,
        source_ip=infra.source_ip,
        asn=infra.asn,
        provider=infra.provider,
        location=infra.location,
        country=infra.country,
        latitude=infra.latitude,
        longitude=infra.longitude,
        domain=infra.domain,
        related_domains=_json.loads(infra.related_domains_json or "[]"),
        related_cases=infra.related_cases,
        threat_score=infra.threat_score,
        infrastructure_type=infra.infrastructure_type,
        threat_feeds=infra.threat_feeds,
    )


@router.get("/{case_id}/graph")
async def get_graph(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Case).where(Case.id == case_id))
    case = result.scalar_one_or_none()
    if not case:
        raise HTTPException(404, "Case not found")

    infra_result = await db.execute(select(InfrastructureRecord).where(InfrastructureRecord.case_id == case_id))
    infra = infra_result.scalar_one_or_none()

    nodes = [
        {
            "id": "case", "node_type": "email", "label": f"Case #{case.number}",
            "sublabel": case.subject[:50], "x": 295, "y": 290,
            "risk_score": case.risk_score, "icon_class": "mail",
            "node_class": "bg-primary text-on-primary shadow-md",
        },
    ]

    if case.sender_domain:
        nodes.append({
            "id": "sender_domain", "node_type": "domain", "label": case.sender_domain,
            "sublabel": "Sender Domain", "x": 100, "y": 120,
            "risk_score": 54, "icon_class": "domain",
            "node_class": "bg-secondary-fixed text-secondary shadow-sm",
        })

    if case.received_ip:
        nodes.append({
            "id": "source_ip", "node_type": "ip", "label": case.received_ip,
            "sublabel": "Origin IP", "x": 480, "y": 110,
            "risk_score": infra.threat_score if infra else 50,
            "location": infra.location if infra else "",
            "provider": infra.provider if infra else "",
            "asn_value": infra.asn if infra else "",
            "icon_class": "router", "badge": "Suspicious" if case.risk_score > 70 else None,
            "node_class": "bg-error-container text-error shadow-md",
        })

    if case.reply_to:
        reply_domain = case.reply_to.split("@")[-1] if "@" in case.reply_to else ""
        if reply_domain:
            nodes.append({
                "id": "reply_domain", "node_type": "domain", "label": reply_domain,
                "sublabel": "Reply-To Target", "x": 100, "y": 410,
                "risk_score": 80, "icon_class": "link_off",
                "node_class": "bg-surface-container-highest text-secondary shadow-sm",
            })

    if case.campaign_id:
        nodes.append({
            "id": "campaign", "node_type": "campaign", "label": "Related Cases",
            "sublabel": f"Campaign {case.campaign_id}", "x": 295, "y": 500,
            "risk_score": 89, "icon_class": "workspaces",
            "node_class": "bg-tertiary-fixed text-tertiary shadow-sm", "badge": "Related",
        })

    edges = []
    if case.sender_domain:
        edges.append({"edge_from": "case", "edge_to": "sender_domain", "label": "header_from", "color": "#6750A4", "edge_type": "dashed"})
    if case.received_ip and case.sender_domain:
        edges.append({"edge_from": "sender_domain", "edge_to": "source_ip", "label": "resolves_to", "color": "#B3261E", "edge_type": "solid"})
    if case.reply_to:
        edges.append({"edge_from": "case", "edge_to": "reply_domain", "label": "reply_to", "color": "#c7c4d8", "edge_type": "dashed"})
    if case.campaign_id:
        edges.append({"edge_from": "case", "edge_to": "campaign", "label": "correlated", "color": "#006e4b", "edge_type": "solid"})

    return {"nodes": nodes, "edges": edges}


@router.get("/{case_id}/campaign", response_model=Optional[CampaignResponse])
async def get_campaign(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Case).where(Case.id == case_id))
    case = result.scalar_one_or_none()
    if not case or not case.campaign_id:
        return None
    from models import CampaignRecord
    camp_result = await db.execute(select(CampaignRecord).where(CampaignRecord.id == case.campaign_id))
    campaign = camp_result.scalar_one_or_none()
    if not campaign:
        return None
    timeline = _json.loads(campaign.timeline_json or "[]")
    return CampaignResponse(
        id=campaign.id,
        name=campaign.name,
        related_emails=campaign.related_emails,
        shared_domains=_json.loads(campaign.shared_domains_json or "[]"),
        shared_infrastructure=_json.loads(campaign.shared_infra_json or "[]"),
        related_cases=campaign.related_cases,
        timeline=[CampaignTimelineEventOut(**t) for t in timeline],
        description=campaign.description,
    )


@router.get("/{case_id}/timeline", response_model=list[TimelineEventResponse])
async def get_timeline(case_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(TimelineEventRecord).where(TimelineEventRecord.case_id == case_id)
        .order_by(TimelineEventRecord.id)
    )
    events = result.scalars().all()
    return [
        TimelineEventResponse(
            id=e.event_id, time=e.time, title=e.title,
            description=e.description, event_type=e.event_type,
            icon=e.icon, details=e.details,
        )
        for e in events
    ]


# ─── Helpers ───────────────────────────────────────────────────

async def _require_evidence(db: AsyncSession, case_id: str) -> EvidenceRecord | None:
    """Return the most recent evidence record for a case, after 404-checking the case."""
    if await evidence_queries.get_case(db, case_id) is None:
        raise HTTPException(404, "Case not found")
    return await evidence_queries.get_latest_evidence(db, case_id)


def _parse_json_list(json_str: str | None) -> list:
    if not json_str:
        return []
    try:
        return json.loads(json_str)
    except Exception:
        return []
